iceoryx_posh/include/iceoryx_posh/roudi/memory/memory_block.hpp

```diff
@@ -52,19 +55,20 @@ class MemoryBlock
     /// @return the alignment of the underlying data.
     virtual uint64_t alignment() const noexcept = 0;
 
+    /// @brief This function provides the pointer to the requested memory.
+    /// @return an optional pointer to a memory block with the requested size and alignment if the memory is available,
+    /// otherwise a cxx::nullopt_t
+    cxx::optional<void*> memory() const noexcept;
+
+  protected:
     /// @brief The MemoryProvider calls this either when MemoryProvider::destroy is called or in its destructor.
     /// @note This function can be called multiple times. Make sure that the implementation can handle this.
     virtual void destroy() noexcept = 0;
 
     /// @brief This function is called once the memory is available and is therefore the earliest possibility to use the
     /// memory.
     /// @param [in] memory pointer to a valid memory block, the same one that the memory() member function would return
-    virtual void memoryAvailable(void* memory) noexcept;
-
-    /// @brief This function provides the pointer to the requested memory.
-    /// @return an optional pointer to a memory block with the requested size and alignment if the memory is available,
-    /// otherwise a cxx::nullopt_t
-    cxx::optional<void*> memory() const noexcept;
+    virtual void memoryAvailable(cxx::not_null<void*> memory) noexcept;
```

Comment:
I understand the purpose of this "callback" but find the name a bit misleading.
It is called automatically at a certain condition so we could set the prefix "on" here.
Maybe something like "onMemoryBlockCreated". But these are only my 2 cents.